File: src/core/point_generator.py
import numpy as np
from src.models.project import Point


def generate_points(
    image_width: int,
    image_height: int,
    count: int,
    distribution: str = "random",
    border: int = 0,
    border_rect: list | None = None,
    border_polygon: list | None = None,
) -> list[Point]:
    """
    Generate points on an image.

    Args:
        image_width: Width of the image in pixels
        image_height: Height of the image in pixels
        count: Number of points to generate
        distribution: 'random', 'stratified', or 'uniform'
        border: Uniform pixel border to exclude (ignored when border_rect or border_polygon is set)
        border_rect: [x_min, y_min, x_max, y_max] from manual click; overrides border
        border_polygon: [[x, y], ...] polygon from manual drawing; overrides border_rect

    Returns:
        List of Point objects
    """
    if border_polygon and len(border_polygon) >= 3:
        coords = _generate_in_polygon(border_polygon, count, distribution)
        return [Point(x=float(x), y=float(y), index=i) for i, (x, y) in enumerate(coords)]

    if border_rect:
        x_min, y_min, x_max, y_max = border_rect
    else:
        x_min = border
        x_max = image_width - border
        y_min = border
        y_max = image_height - border

    if x_min >= x_max or y_min >= y_max:
        raise ValueError("Border exclusion too large for image size.")

    if distribution == "random":
        coords = _random_points(x_min, x_max, y_min, y_max, count)
    elif distribution == "stratified":
        coords = _stratified_points(x_min, x_max, y_min, y_max, count)
    elif distribution == "uniform":
        coords = _uniform_grid_points(x_min, x_max, y_min, y_max, count)
    else:
        raise ValueError(f"Unknown distribution: {distribution}")

    return [Point(x=float(x), y=float(y), index=i) for i, (x, y) in enumerate(coords)]


def _point_in_polygon(x: float, y: float, polygon: list) -> bool:
    """Ray-casting point-in-polygon test."""
    n = len(polygon)
    inside = False
    j = n - 1
    for i in range(n):
        xi, yi = float(polygon[i][0]), float(polygon[i][1])
        xj, yj = float(polygon[j][0]), float(polygon[j][1])
        if ((yi > y) != (yj > y)) and (x < (xj - xi) * (y - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside


def _generate_in_polygon(polygon: list, count: int, distribution: str) -> list[tuple]:
    poly = np.array(polygon, dtype=float)
    x_min, y_min = float(poly[:, 0].min()), float(poly[:, 1].min())
    x_max, y_max = float(poly[:, 0].max()), float(poly[:, 1].max())

    if distribution == "uniform":
        density = max(count * 9, 900)
        cols = int(np.ceil(np.sqrt(density)))
        rows = int(np.ceil(density / cols))
        xs = np.linspace(x_min, x_max, cols)
        ys = np.linspace(y_min, y_max, rows)
        xv, yv = np.meshgrid(xs, ys)
        candidates = list(zip(xv.ravel().tolist(), yv.ravel().tolist()))
        inside = [(x, y) for x, y in candidates if _point_in_polygon(x, y, polygon)]
        if len(inside) >= count:
            step = max(1, len(inside) // count)
            return inside[::step][:count]
        return inside

    # random / stratified: rejection sampling
    coords: list[tuple] = []
    batch = max(count * 4, 200)
    max_iters = 50
    for _ in range(max_iters):
        if len(coords) >= count:
            break
        if distribution == "stratified":
            xs_b = np.random.uniform(x_min, x_max, batch)
            ys_b = np.random.uniform(y_min, y_max, batch)
        else:
            xs_b = np.random.uniform(x_min, x_max, batch)
            ys_b = np.random.uniform(y_min, y_max, batch)
        for x, y in zip(xs_b.tolist(), ys_b.tolist()):
            if len(coords) >= count:
                break
            if _point_in_polygon(x, y, polygon):
                coords.append((x, y))
    return coords[:count]


def _random_points(x_min, x_max, y_min, y_max, count):
    xs = np.random.uniform(x_min, x_max, count)
    ys = np.random.uniform(y_min, y_max, count)
    return list(zip(xs, ys))


def _stratified_points(x_min, x_max, y_min, y_max, count):
    """Stratified random sampling — divides image into grid cells, one point per cell."""
    cols = int(np.ceil(np.sqrt(count)))
    rows = int(np.ceil(count / cols))

    cell_w = (x_max - x_min) / cols
    cell_h = (y_max - y_min) / rows

    coords = []
    for row in range(rows):
        for col in range(cols):
            if len(coords) >= count:
                break
            x = x_min + col * cell_w + np.random.uniform(0, cell_w)
            y = y_min + row * cell_h + np.random.uniform(0, cell_h)
            x = min(x, x_max)
            y = min(y, y_max)
            coords.append((x, y))

    return coords[:count]


def _uniform_grid_points(x_min, x_max, y_min, y_max, count):
    """Uniform grid — evenly spaced points across the image."""
    cols = int(np.ceil(np.sqrt(count)))
    rows = int(np.ceil(count / cols))

    xs = np.linspace(x_min, x_max, cols)
    ys = np.linspace(y_min, y_max, rows)

    coords = [(x, y) for y in ys for x in xs]
    return coords[:count]

File: rust/src/core/point_generator.rs
use anyhow::{bail, Result};
use rand::RngExt;

use crate::models::Point;

/// Generate points on an image.
///
/// * `distribution` - "random", "stratified", or "uniform"
/// * `border` - uniform pixel border to exclude (ignored when `border_rect` or
///   `border_polygon` is set)
/// * `border_rect` - `[x_min, y_min, x_max, y_max]` from a manual click; overrides `border`
/// * `border_polygon` - `[[x, y], ...]` from manual drawing; overrides `border_rect`
#[allow(clippy::too_many_arguments)]
pub fn generate_points(
    image_width: f64,
    image_height: f64,
    count: usize,
    distribution: &str,
    border: f64,
    border_rect: Option<&[f64]>,
    border_polygon: Option<&[Vec<f64>]>,
) -> Result<Vec<Point>> {
    if let Some(polygon) = border_polygon {
        if polygon.len() >= 3 {
            let coords = generate_in_polygon(polygon, count, distribution);
            return Ok(coords
                .into_iter()
                .enumerate()
                .map(|(i, (x, y))| Point { x, y, index: i as i64, label: None, category: None })
                .collect());
        }
    }

    let (x_min, y_min, x_max, y_max) = if let Some(rect) = border_rect {
        (rect[0], rect[1], rect[2], rect[3])
    } else {
        (border, border, image_width - border, image_height - border)
    };

    if x_min >= x_max || y_min >= y_max {
        bail!("Border exclusion too large for image size.");
    }

    let coords = match distribution {
        "random" => random_points(x_min, x_max, y_min, y_max, count),
        "stratified" => stratified_points(x_min, x_max, y_min, y_max, count),
        "uniform" => uniform_grid_points(x_min, x_max, y_min, y_max, count),
        other => bail!("Unknown distribution: {other}"),
    };

    Ok(coords
        .into_iter()
        .enumerate()
        .map(|(i, (x, y))| Point { x, y, index: i as i64, label: None, category: None })
        .collect())
}

/// Ray-casting point-in-polygon test.
fn point_in_polygon(x: f64, y: f64, polygon: &[Vec<f64>]) -> bool {
    let n = polygon.len();
    let mut inside = false;
    let mut j = n - 1;
    for i in 0..n {
        let (xi, yi) = (polygon[i][0], polygon[i][1]);
        let (xj, yj) = (polygon[j][0], polygon[j][1]);
        if (yi > y) != (yj > y) && x < (xj - xi) * (y - yi) / (yj - yi) + xi {
            inside = !inside;
        }
        j = i;
    }
    inside
}

fn generate_in_polygon(polygon: &[Vec<f64>], count: usize, distribution: &str) -> Vec<(f64, f64)> {
    let x_min = polygon.iter().map(|p| p[0]).fold(f64::INFINITY, f64::min);
    let x_max = polygon.iter().map(|p| p[0]).fold(f64::NEG_INFINITY, f64::max);
    let y_min = polygon.iter().map(|p| p[1]).fold(f64::INFINITY, f64::min);
    let y_max = polygon.iter().map(|p| p[1]).fold(f64::NEG_INFINITY, f64::max);

    if distribution == "uniform" {
        let density = (count * 9).max(900);
        let cols = (density as f64).sqrt().ceil() as usize;
        let rows = (density as f64 / cols as f64).ceil() as usize;
        let xs = linspace(x_min, x_max, cols);
        let ys = linspace(y_min, y_max, rows);

        let mut inside: Vec<(f64, f64)> = Vec::new();
        for &y in &ys {
            for &x in &xs {
                if point_in_polygon(x, y, polygon) {
                    inside.push((x, y));
                }
            }
        }

        if inside.len() >= count && count > 0 {
            let step = (inside.len() / count).max(1);
            return inside.into_iter().step_by(step).take(count).collect();
        }
        return inside;
    }

    // random / stratified: rejection sampling (matches the Python implementation,
    // which does not actually stratify this branch either).
    let mut coords: Vec<(f64, f64)> = Vec::new();
    let batch = (count * 4).max(200);
    let mut rng = rand::rng();
    for _ in 0..50 {
        if coords.len() >= count {
            break;
        }
        for _ in 0..batch {
            if coords.len() >= count {
                break;
            }
            let x = rng.random_range(x_min..=x_max);
            let y = rng.random_range(y_min..=y_max);
            if point_in_polygon(x, y, polygon) {
                coords.push((x, y));
            }
        }
    }
    coords.truncate(count);
    coords
}

fn random_points(x_min: f64, x_max: f64, y_min: f64, y_max: f64, count: usize) -> Vec<(f64, f64)> {
    let mut rng = rand::rng();
    (0..count)
        .map(|_| (rng.random_range(x_min..=x_max), rng.random_range(y_min..=y_max)))
        .collect()
}

/// Stratified random sampling — divides the image into grid cells, one point per cell.
fn stratified_points(x_min: f64, x_max: f64, y_min: f64, y_max: f64, count: usize) -> Vec<(f64, f64)> {
    if count == 0 {
        return Vec::new();
    }
    let cols = (count as f64).sqrt().ceil() as usize;
    let rows = (count as f64 / cols as f64).ceil() as usize;

    let cell_w = (x_max - x_min) / cols as f64;
    let cell_h = (y_max - y_min) / rows as f64;

    let mut rng = rand::rng();
    let mut coords = Vec::with_capacity(count);
    'outer: for row in 0..rows {
        for col in 0..cols {
            if coords.len() >= count {
                break 'outer;
            }
            let x = (x_min + col as f64 * cell_w + rng.random_range(0.0..=cell_w)).min(x_max);
            let y = (y_min + row as f64 * cell_h + rng.random_range(0.0..=cell_h)).min(y_max);
            coords.push((x, y));
        }
    }
    coords.truncate(count);
    coords
}

/// Uniform grid — evenly spaced points across the image.
fn uniform_grid_points(x_min: f64, x_max: f64, y_min: f64, y_max: f64, count: usize) -> Vec<(f64, f64)> {
    if count == 0 {
        return Vec::new();
    }
    let cols = (count as f64).sqrt().ceil() as usize;
    let rows = (count as f64 / cols as f64).ceil() as usize;

    let xs = linspace(x_min, x_max, cols);
    let ys = linspace(y_min, y_max, rows);

    let mut coords = Vec::with_capacity(rows * cols);
    for &y in &ys {
        for &x in &xs {
            coords.push((x, y));
        }
    }
    coords.truncate(count);
    coords
}

/// Mirrors `numpy.linspace(start, stop, num)`: `num` evenly spaced points
/// over `[start, stop]` inclusive, with the last point forced to `stop`.
fn linspace(start: f64, stop: f64, num: usize) -> Vec<f64> {
    if num == 0 {
        return Vec::new();
    }
    if num == 1 {
        return vec![start];
    }
    let step = (stop - start) / (num - 1) as f64;
    let mut values: Vec<f64> = (0..num).map(|i| start + i as f64 * step).collect();
    values[num - 1] = stop;
    values
}
